File: leetcode/lc-us/cpp/203.cpp
struct ListNode {
    int val;
    ListNode *next;

    ListNode() : val(0), next(nullptr) {}

    ListNode(int x) : val(x), next(nullptr) {}

    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode *removeElements(ListNode *head, int val) {
        auto dummy = new ListNode(0, head), node = dummy;
        while (node && node->next) {
            while (node->next && node->next->val == val) {
                node->next = node->next->next;
            }
            node = node->next;
        }
        return dummy->next;
    }
};

File: leetcode/lc-us/rust/203.rs
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode {
            next: None,
            val,
        }
    }
}

impl Solution {
    pub fn remove_elements(head: Option<Box<ListNode>>, val: i32) -> Option<Box<ListNode>> {
        match head {
            None => None,
            Some(mut node) => {
                let right = Self::remove_elements(node.next, val);
                if node.val == val {
                    right
                } else {
                    node.next = right;
                    Some(node)
                }
            }
        }
    }
}